### PR TITLE
support multi_scale_deform_attn trt plugin

### PR DESCRIPTION
**Pull request by cxiang26**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

This trt plugin refers to [mmcv](https://github.com/open-mmlab/mmcv/blob/2.x/mmcv/ops/csrc/common/cuda/points_in_boxes_cuda_kernel.cuh) and [tensorrt](https://github.com/NVIDIA/TensorRT/blob/main/plugin/multiscaleDeformableAttnPlugin/README.md)

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by AllentDan — COMMENTED** [review]
There are some works to do for you:
1. add symbolic rewriting of this op in MMDeploy to export ONNX.
2. add corresponding unit test
3. update [doc description for customized ops](https://github.com/open-mmlab/mmdeploy/blob/dev-1.x/docs/en/06-custom-ops/tensorrt.md)

BTW, what model does this op work for?

**Comment by AllentDan** [comment]
Please fix the lint issue. You can install pre-commit hook on your localhost like following:
```shell
pip install -U pre-commit
cd mmdeploy
pre-commit install
```
You can run `pre-commit run --all-file` to check whether there are lint errors.

As for cscr files, you may try `clang-format`.

**Comment by cxiang26** [comment]
> 

this plugin is for [bevformer](https://github.com/fundamentalvision/BEVFormer/blob/master/projects/mmdet3d_plugin/bevformer/modules/spatial_cross_attention.py)

File: csrc/mmdeploy/backend_ops/tensorrt/multi_scale_deform_attn/trt_ms_deform_attn_kernel.hpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) OpenMMLab. All rights reserved
+#ifndef TRT_MS_DEFORM_ATTN_KERNEL_HPP
+#define TRT_MS_DEFORM_ATTN_KERNEL_HPP
+#include <cublas_v2.h>
+#include <cuda_runtime.h>
+
+int32_t ms_deform_attn_cuda_forward(const float* value, const int32_t* spatialShapes,
```
Comment (grimoire) [inline csrc/mmdeploy/backend_ops/tensorrt/multi_scale_deform_attn/trt_ms_deform_attn_kernel.hpp:7]: Why don't we create a template here?

**Comment by grimoire** [comment]
If your code is modified from other repo, please add some comment at the head of the file like https://github.com/open-mmlab/mmdeploy/blob/master/csrc/mmdeploy/backend_ops/torchscript/optimizer/ir/subgraph_matcher.cpp .
Any please provide the unit test so we can test the op.

File: csrc/mmdeploy/backend_ops/tensorrt/multi_scale_deform_attn/trt_ms_deform_attn_kernel.hpp (hunk 1 above)
Comment (cxiang26) [reply csrc/mmdeploy/backend_ops/tensorrt/multi_scale_deform_attn/trt_ms_deform_attn_kernel.hpp]: fixed

**Review by grimoire — COMMENTED** [review]
Unit test failed with log: 
```
FAILED tests/test_ops/test_ops.py::test_multi_scale_deformable_attn[backend0] - RuntimeError: CUDA error: an illegal memory access was encountered
```

**Comment by grimoire** [comment]
result mismatch
```
E               AssertionError: Tensor-likes are not close!
E               
E               Mismatched elements: 16384 / 16384 (100.0%)
E               Greatest absolute difference: 11.818938732147217 at index (0, 4, 130) (up to 1e-05 allowed)
E               Greatest relative difference: 0.6782912324431448 at index (1, 24, 59) (up to 0.001 allowed)
```

**Comment by cxiang26** [comment]
> result mismatch
> 
> ```
> E               AssertionError: Tensor-likes are not close!
> E               
> E               Mismatched elements: 16384 / 16384 (100.0%)
> E               Greatest absolute difference: 11.818938732147217 at index (0, 4, 130) (up to 1e-05 allowed)
> E               Greatest relative difference: 0.6782912324431448 at index (1, 24, 59) (up to 0.001 allowed)
> ```
The mismatch is caused by input[1] & input[2] point to the same address in [plugin.cpp](https://github.com/open-mmlab/mmdeploy/pull/1844/files#diff-5dbef9c4fb41e37c032bb787d8378f584ce3ababa349479edb34aeb7a94708a2R97), input[1] and input [2] should difference
I have doubled check this plugin, but I can't find the reason resulting in this bug
BTW, the deployed result with this plugin can obtain the right result in my project

**Comment by grimoire** [comment]
Sorry, It is my mistake.
https://github.com/open-mmlab/mmdeploy/blob/34c68663b6c1865b702247dc48d99781b8bd748e/mmdeploy/backend/tensorrt/wrapper.py#L155
input_tensor will be released after the loop, that leads to the error.

You can initialize the inputs before the for loop:
```python

        inputs = dict((name, data.contiguous().int() if data.dtype ==
                       torch.long else data.contiguous())
                      for name, data in inputs.items())
        for input_name, input_tensor in inputs.items():
             ...
```

Could you please fix it for me in this PR.

**Review by grimoire — APPROVED** [review]
LGTM

**Review by AllentDan — APPROVED** [review]
LGTM

**Pull request merged**